File: Core/actor/toggleactor.cpp
#include "toggleactor.h"

#include <QDebug>


ToggleActor::ToggleActor(ValueGroup *valueGroup, QString id, QObject *parent)
    : ActorBase(valueGroup, id, VALUE_TYPE::VALTYPE_VIRTUAL_ACTOR, QVariant::Bool, parent)
{
    m_alwaysEmit = false;
}

bool ToggleActor::cmdSupported(actor::ACTOR_CMDS cmd) {
    switch(cmd) {
    case actor::ACTOR_CMD_ON:
    case actor::ACTOR_CMD_OFF:
    case actor::ACTOR_CMD_TOGGLE:
        return true;
    default:
        return false;
    }
}

QVariant ToggleActor::_updateValue(QVariant newValue) {
    iDebug() << Q_FUNC_INFO << newValue;

    if (newValue.canConvert(QVariant::Bool)) {
        return QVariant::fromValue(newValue);
    } else {
        return QVariant();
    }
}

bool ToggleActor::isAsync() {
    return false;
}

void ToggleActor::_triggerCmd(actor::ACTOR_CMDS cmd) {
    switch(cmd) {
    case actor::ACTOR_CMD_ON:
        updateValue(true);
        break;
    case actor::ACTOR_CMD_OFF:
        updateValue(false);
        break;
    case actor::ACTOR_CMD_TOGGLE:
        updateValue(!rawValue().toBool());
        break;
    default:
        break;
    }
}

File: AndroidApp/OshCore/src/main/java/com/osh/actor/ToggleActor.java
package com.osh.actor;

import com.osh.value.ValueGroup;
import com.osh.value.ValueType;

public class ToggleActor extends ActorBase<ToggleActor, Boolean> {
    public ToggleActor(ValueGroup valueGroup, String id) {
        super(valueGroup, id, ValueType.VT_VIRTUAL_ACTOR);
    }

    @Override
    public boolean cmdSupported(ActorCmds cmd) {
        return cmd == ActorCmds.ACTOR_CMD_TOGGLE;
    }

    @Override
    protected void _triggerCmd(ActorCmds cmd) {

    }

    @Override
    protected Boolean _updateValue(Object newValue) {
        if (newValue instanceof Boolean) {
            return (Boolean) newValue;
        } else {
            return null;
        }
    }
}
